lib/functoria/impl.ml

```diff
@@ -53,84 +82,316 @@ let main ?packages ?packages_v ?keys ?extra_deps module_name ty =
+    | Nil -> ()
+    | Cons (h, t) -> Fmt.pf ppf "%a,@ %a" pp h pp_tl t
 
 and pp_abstract ppf (Abstract i) = pp ppf i
 
 let rec hash : type a. a t -> int = function
-  | Dev c -> Device.hash c
-  | App { f; x } -> Hashtbl.hash (hash f, hash x)
-  | If (cond, t, e) -> Hashtbl.hash (cond, hash t, hash e)
-
-let hash_abstract (Abstract x) = hash x
+  | Dev { dev; args; deps } ->
+    Hashtbl.hash
+      (`Dev, Device.hash dev, hash_tl args, List.map hash_abstract deps)
+  | App { f; args } ->
+    Hashtbl.hash
+      (`App, hash f, hash_tl args)
+  | If {cond; branches; default} ->
+    Hashtbl.hash
+      (`If, cond, List.map (fun (p,t) -> Hashtbl.hash (p, hash t)) branches, hash default)
+and hash_abstract (Abstract x) = hash x
+and hash_tl : type a v. (a,v) tl -> int = fun x -> match x with
+  | Nil -> Hashtbl.hash `Nil
+  | Cons (h, t) ->
+    Hashtbl.hash (`Cons, hash h, hash_tl t)
 
+type ex = Ex : 'a -> ex
 let rec equal : type t1 t2. t1 t -> t2 t -> bool =
- fun x y ->
-  match (x, y) with
-  | Dev c, Dev c' -> Device.equal c c'
-  | App a, App b -> equal a.f b.f && equal a.x b.x
-  | If (cond1, t1, e1), If (cond2, t2, e2) ->
-      (* Key.value is a functional value (it contains a closure for eval).
-         There is no prettier way than physical equality. *)
-      cond1 == cond2 && equal t1 t2 && equal e1 e2
+  fun x y ->
+  match x, y with
+  | Dev c, Dev c' ->
+    Device.equal c.dev c'.dev &&
+    equal_tl c.args c'.args &&
+    List.for_all2 equal_abstract c.deps c'.deps
+  | App a, App b ->
+    equal a.f b.f && equal_tl a.args b.args
+  | If x1, If x2 ->
+    (* Key.value is a functional value (it contains a closure for eval).
+       There is no prettier way than physical equality. *)
+    Obj.repr x1.cond == Obj.repr x2.cond &&
+    List.for_all2
+      (fun (p1,t1) (p2,t2) -> Ex p1 = Ex p2 && equal t1 t2)
+      x1.branches x2.branches &&
+    equal x1.default x2.default
+  | _ -> false
+and equal_abstract (Abstract x) (Abstract y) = equal x y
+and equal_tl : type t1 t2 v1 v2. (t1,v1) tl -> (t2,v2) tl -> bool =
+  fun x y ->
+  match x, y with
+  | Nil, Nil -> true
+  | Cons (h1, t1), Cons (h2, t2) ->
+    equal h1 h2 && equal_tl t1 t2
   | _ -> false
-
-let equal_abtract (Abstract a) (Abstract b) = equal a b
 
 module Tbl = Hashtbl.Make (struct
   type t = abstract
 
   let hash = hash_abstract
 
-  let equal = equal_abtract
+  let equal = equal_abstract
 end)
 
-type 'b f_dev = { f : 'a. ('a, abstract) Device.t -> 'b }
+(* Safe, but annoying to code safely, so we cheat *)
+module HC : sig
+  type tbl
+  val create : unit -> tbl
+  val add : tbl -> 'a t -> 'a t -> unit
+  val get : tbl -> 'a t -> 'a t option
+end = struct
+  type tbl = abstract Tbl.t
+  let create () = Tbl.create 50
+  let add tbl a b = Tbl.add tbl (abstract a) (abstract b)
+  let get (type a) tbl (a : a t) =
+    if Tbl.mem tbl @@ abstract a then
+      let Abstract b = Tbl.find tbl (abstract a) in
+      Some (Obj.magic b : a t)
+    else
+      None
+end
```

Comment:
This is rather dirty. If anyone has a proposal to make that safe in a concise manner. I'm all ears. So far my plan is to use equality witnesses.

Reply:
Please use equality witnesses 👍

Reply:
Done!